### ISSUE TITLE
Misbehaving with GreaseWeazle

### ISSUE THREAD

**Issue by drdpj** [davidgiven/fluxengine#360]
Greaseweazle 0.31, using an F4

I've two 5.25" drives connected, a 40track as drive 1, and 80 (1.2mb, jumpered for 300rpm) as drive 0.

fluxengine analyse driveresponse --cylinder 0 --min-interval-us=0 --max-interval-us=30 --interval-step-us=.1 --write-img=driveresponse.png --usb.greaseweazle=COM3 -d drive:1

This is a bit hit and miss, sometimes I just get the drive clicking whilst "Measuring rotational speed", sometimes I get this:
"Measuring rotational speed...
Interval 0.00: failed
Interval 0.08: failed
Interval 0.17: failed
Interval 0.25: failed
Interval 0.33: failed
Interval 0.50: failed
Interval 0.58: failed
Interval 0.67: failed
Interval 0.75: failed
Interval 0.83: failed
Interval 1.00: failed
Interval 1.08: failed
Interval 1.17: failed
Interval 1.25: failed
Interval 1.33: failed
Interval 1.50: failed
Interval 1.58: failed
Interval 1.67: failed
Interval 1.75: failed
Interval 1.83: failed
Interval 2.00: Error: failed to send command: No error"

Once I got: "
Interval 1.75: failed
Interval 1.83: failed
Interval 2.00: failed
Interval 2.08: 8.7368
Interval 2.17: 2.1666
Interval 2.25: 2.2499
Interval 2.33: 2.3333
Interval 2.50: 2.4999
Interval 2.58: 2.5833
Interval 2.67: 2.6667
Interval 2.75: 2.7499
Interval 2.83: 2.8333
Interval 3.00: 3.0001
Interval 3.08: 3.0835
Interval 3.17: 3.1666
Interval 3.25: 3.2498
Interval 3.33: 3.3332
Interval 3.50: Error: GreaseWeazle error: Underflow"

If I try and use drive 0, the activity light goes on but the motor doesn't.

**Comment by davidgiven** [comment]
I have an F7 rather than an F4, but it shouldn't make a difference? The last error looks like it's running out of bandwidth, so do make sure it's plugged into a USB 2 port rather than a USB 1 port.

FWIW I do see from the last set of errors that 2.08us pulse intervals are not being reported correctly by the drive (it's getting back 8.7us pulses instead). That probably indicates the minimum pulse width supported by the drive.

**Comment by drdpj** [comment]
No, it shouldn't - actually, mine is an F730 now I look closely :D It's
actually plugged into a USB3 port, so there shouldn't be any issues there.
Any ideas what's going on with drive 0?

I'm happy to debug etc, etc.

**Comment by tdaede** [comment]
FWIW I can reproduce the driveresponse issue for pulses shorter than 2us on a greaseweazle v4.

**Comment by davidgiven** [comment]
Just to be clear, the drives work fine when using GreaseWeazle's own tools?

Also, actually FluxEngine runs out of USB bandwidth writing pulses closer than about 1.8us. (The shorter the pulse intervals, the more data is required.) So there's a hard-coded limit of 2.0us below which it just refuses to do anything. The GreaseWeazle is a USB 3 device and works intermittently down to about 1us, but my drive starts to fail at 1.2us.

I don't have any Victor 9000 disks but I've got an image here which I think has been read in a 300 rpm drive, and the pulse interval of the outer tracks is about 1.45us. I can write that with the GreaseWeazle but it fails miserably on the FluxEngine.

**Comment by drdpj** [comment]
Yes, no problem with the drives at all with the greaseweazle software -
drives 0 and 1 happily co-exist. I've just tried a few different drives, I
can't get below about 2.2us to read reliably with any of my 80 track 5.25s,
which isn't going to be much cop for the victor discs I fear. That's TEACs
and Mitsubishis and NECs... I've got a couple more to try tomorrow.

**Comment by davidgiven** [comment]
What GreaseWeazle firmware do you have, BTW?

**Comment by drdpj** [comment]
0.31 currently

**Comment by tdaede** [comment]
The Greaseweazle V4 is still only USB 1.1 and I can't get more than about 8Msps of bandwidth out of it, which could be the problem.

(in my case I'm always writing pulses 2us or longer so I've had no problems introduced by this)

You might also want to check the state of pin 2 during writing, it should be high for HD disks.

**Comment by davidgiven** [comment]
FWIW, here's what I get on my 1.2MB drive via a GreaseWeazle. You can see the smear in the bottom right from spurious pulses as the interval gets too big, and this really weird repeated pattern on the left when the pulses are too short. The good range seems to be about 1.1us to about 7us.

![driveresponse](https://user-images.githubusercontent.com/100638/144724944-b92474a4-f8e2-401e-9778-20be459bd700.png)

**Comment by tdaede** [comment]
If it's helpful, I've produced a bunch of frequency responses for all of my floppy drives on hand. These were all produced with the Greaseweazle V4, so none of them go below 2.00us.

https://docs.google.com/spreadsheets/d/1EG5jYsqBiQIu0XConvSuIooGkA_WonFZVIxuY0oHloE/edit#gid=0
